src/libinjector/linux/linux_injector.c

```diff
@@ -0,0 +1,225 @@
+#include "linux_utils.h"
+#include "linux_debug.h"
+#include "methods/linux_shellcode.h"
+
+bool check_userspace_int3_trap(injector_t injector, drakvuf_trap_info_t* info)
+{
+    // check CPL
+    unsigned long int CPL = (info->regs->cs_sel & 3);
+    PRINT_DEBUG("CPL 0x%lx\n", CPL);
+
+    if ( CPL != 0)
+    {
+        PRINT_DEBUG("Inside INT3 userspace\n");
+    }
+    else
+    {
+        PRINT_DEBUG("INT3 received but CPL is not 0x3\n");
+        return false;
+    }
+
+    if ( info->proc_data.pid != injector->target_pid )
+    {
+        PRINT_DEBUG("INT3 received but '%s' PID (%u) doesn't match target process (%u)\n",
+            info->proc_data.name, info->proc_data.pid, injector->target_pid);
+        return false;
+    }
+
+    if (info->regs->rip != info->trap->breakpoint.addr)
+    {
+        PRINT_DEBUG("INT3 received but BP_ADDR (%lx) doesn't match RIP (%lx)",
+            info->trap->breakpoint.addr, info->regs->rip);
+        assert(false);
+    }
+
+    if (injector->target_tid && (uint32_t)info->proc_data.tid != injector->target_tid)
+    {
+        PRINT_DEBUG("INT3 received but '%s' TID (%u) doesn't match target process (%u)\n",
+            info->proc_data.name, info->proc_data.tid, injector->target_tid);
+        return false;
+    }
+
+    return true;
+}
+
+static event_response_t injector_int3_userspace_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = info->trap->data;
+
+    PRINT_DEBUG("INT3 Callback @ 0x%lx. CR3 0x%lx. vcpu %i. TID %u\n",
+        info->regs->rip, info->regs->cr3, info->vcpu, info->proc_data.tid);
+
+    if (!check_userspace_int3_trap(injector, info))
+        return VMI_EVENT_RESPONSE_NONE;
+
+    // reset the override on every run
+    injector->step_override = false;
+
+    event_response_t event;
+    switch (injector->method)
+    {
+        case INJECT_METHOD_SHELLCODE:
+        {
+            event = handle_shellcode(drakvuf, info);
+            break;
+        }
+        default:
+        {
+            PRINT_DEBUG("Should not be here\n");
+            assert(false);
+        }
+    }
```

Comment:
Looks like an if statement to me. Don't overcomplicate code "for the future.

Reply:
it really does not look "overcomplicated" to me. if possible, can we please avoid this change?

Reply:
I think we can leave it, it will become clear whether it is overcomplicated once you add the read/write file part.